File: app/src/main/java/ua/meugen/android/pdfcomposer/ui/activities/base/fragment/presenter/BasePresenter.java
package ua.meugen.android.pdfcomposer.ui.activities.base.fragment.presenter;

import io.reactivex.disposables.CompositeDisposable;
import ua.meugen.android.pdfcomposer.ui.activities.base.fragment.state.MvpState;

public class BasePresenter<S extends MvpState> implements MvpPresenter<S> {

    private CompositeDisposable compositeDisposable;

    @Override
    public void onSaveState(final S state) {}

    @Override
    public void onRestoreState(final S state) {}

    @Override
    public void clear() {
        if (compositeDisposable != null) {
            compositeDisposable.clear();
        }
        compositeDisposable = null;
    }

    protected final CompositeDisposable getCompositeDisposable() {
        if (compositeDisposable == null) {
            compositeDisposable = new CompositeDisposable();
        }
        return compositeDisposable;
    }
}

File: app/src/main/java/ua/meugen/android/pdfcomposer/ui/activities/base/fragment/presenter/BasePresenter.kt
package ua.meugen.android.pdfcomposer.ui.activities.base.fragment.presenter

import io.reactivex.disposables.CompositeDisposable
import ua.meugen.android.pdfcomposer.ui.activities.base.fragment.state.MvpState

open class BasePresenter<S : MvpState> : MvpPresenter<S> {

    private var compositeDisposable: CompositeDisposable? = null

    override fun onSaveState(state: S) {}

    override fun onRestoreState(state: S) {}

    override fun clear() {
        compositeDisposable?.clear()
        compositeDisposable = null
    }

    protected fun getCompositeDisposable(): CompositeDisposable {
        compositeDisposable?.let { return it }
        return CompositeDisposable().also { compositeDisposable = it }
    }
}
